crates/biome_rowan/src/ast/batch.rs

```diff
@@ -348,89 +343,189 @@ where
-                return root;
+            }
+            // If parent is None, we reached the document root
+            else {
+                let optional_text_range_and_edit = if with_text_range_and_edit {
+                    // The root of batch mutation is not necessarily
+                    // the document root in some rule actions,
+                    // so we need to find the actual document root
+                    let mut document_root = root;
+                    while let Some(parent) = document_root.parent() {
+                        document_root = parent;
+                    }
+
+                    if curr_is_from_action {
+                        text_mutation_list = vec![(
+                            document_root.text_range(),
+                            Some(
+                                curr_new_node
+                                    .as_ref()
+                                    .map_or(String::new(), |n| n.to_string()),
+                            ),
+                        )];
+                    }
+
+                    // Build text range and text edit from the text mutation list
+                    let root_string = document_root.to_string();
+                    let mut text_range = TextRange::default();
+                    let mut text_edit_builder = TextEditBuilder::default();
+
+                    let mut pointer: usize = 0;
+                    for (deleted_text_range, optional_inserted_text) in text_mutation_list {
+                        if let (Ok(range_start), Ok(range_end)) = (
+                            usize::try_from(u32::from(deleted_text_range.start())),
+                            usize::try_from(u32::from(deleted_text_range.end())),
+                        ) {
+                            text_range = text_range.cover(deleted_text_range);
+                            if range_start > pointer {
+                                text_edit_builder.equal(&root_string[pointer..range_start]);
+                            }
+
+                            let old = &root_string[range_start..range_end];
+                            let new = &optional_inserted_text.map_or(String::new(), |t| t);
+
+                            text_edit_builder.with_unicode_words_diff(old, new);
+
+                            pointer = range_end;
+                        }
+                    }
+                    let end_pos = root_string.len();
+                    if end_pos > pointer {
+                        text_edit_builder.equal(&root_string[pointer..end_pos]);
+                    }
+
+                    let text_edit = text_edit_builder.finish();
+
+                    Some((text_range, text_edit))
+                } else {
+                    None
+                };
+
+                return (
+                    curr_new_node
+                        .expect("new_node")
+                        .into_node()
+                        .expect("expected root to be a node and not a token"),
```

Comment:
Can we add a `// SAFETY` comment that explains how it is safe to use `expect()`?